### PR TITLE
Chat with any compatible targets in Target MVD

### PR DESCRIPTION
**Pull request by afmicc**
#### Trello board reference:

* [As a user I should be able to send a message in a conversation with any of my compatible targets so I can talk about our common interests](https://trello.com/c/16IkYJiJ)

---

#### Description:

* Add chat models and channel
* Create chat room after create compatible target

### CODE REVIEW

File: app/channels/application_cable/chat_room_channel.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+    def unsubscribed
+      # Any cleanup needed when channel is unsubscribed
+    end
+
+    def send_message(data)
+      ActiveRecord::Base.transaction do
+        body = data['message']
+        chat_room_id = data['chat_room_id']
+        message = current_user.messages.create! body: body,
```
Comment (gabrielbursztein2) [inline app/channels/application_cable/chat_room_channel.rb]: Add `()`

File: app/channels/application_cable/chat_room_channel.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+      transmit error: I18n.t('api.errors.not_found')
+      reject
+    end
+
+    def unsubscribed
+      # Any cleanup needed when channel is unsubscribed
+    end
+
+    def send_message(data)
+      ActiveRecord::Base.transaction do
```
Comment (gabrielbursztein2) [inline app/channels/application_cable/chat_room_channel.rb:15]: Why is this wrapped in a transaction? There is no database call.

File: app/channels/application_cable/connection.rb [hunk 3]
```diff
@@ -1,4 +1,34 @@
...
+      logger.add_tags 'ActionCable', current_user.email
+    end
+
+    protected
+
+    def find_verified_user
+      params = header_params
+      user = User.find_by_uid(params[:uid])
+
+      return reject_unauthorized_connection if
```
Comment (gabrielbursztein2) [inline app/channels/application_cable/connection.rb]: No need to return, just call `reject_unauthorized_connection`

File: app/controllers/api/v1/chat_rooms_controller.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+module Api
+  module V1
+    class ChatRoomsController < ApiController
+      def index
+        @chat_rooms = current_user.chat_rooms
+      end
+
+      def show
+        @chat_room = current_user.chat_rooms.includes(:messages).find_by!(id: params[:id])
```
Comment (gabrielbursztein2) [inline app/controllers/api/v1/chat_rooms_controller.rb]: There is no need to pre include messages when there is only one chat_room.
Also, just use `find` when searching by `id`

File: app/models/chat_room.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+#  title         :string           not null
+#  user_owner_id :bigint           not null
+#  user_guest_id :bigint           not null
+#  created_at    :datetime         not null
+#  updated_at    :datetime         not null
+#
+
+class ChatRoom < ActiveRecord::Base
+  belongs_to :user_owner, class_name: :User, foreign_key: 'user_owner_id'
+  belongs_to :user_guest, class_name: :User, foreign_key: 'user_guest_id'
```
Comment (gabrielbursztein2) [inline app/models/chat_room.rb:15]: I would prefer a `has_many :users` relation. This way you avoid `own_chat_rooms ` and `guest_chat_rooms` in the user model.

File: app/models/target.rb [hunk 6]
```diff
@@ -63,9 +63,23 @@ def validate_target_limit
...
+    targets = near_targets
+    ActiveRecord::Base.transaction do
+      create_chat_rooms_compatibles(targets)
+      notify_compatible(targets)
+    end
+  end
+
+  def create_chat_rooms_compatibles(targets)
+    targets.each do |target|
+      user.own_chat_rooms.create! title: "#{title} - #{target.title}",
```
Comment (gabrielbursztein2) [inline app/models/target.rb:65]: `()`

File: spec/channels/chat_room_channel_spec.rb [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+    it 'is expected a successful subscription' do
+      expect(subscription).to be_confirmed
+      expect(subscription).to have_stream_for(chat_room)
+    end
+
+    context 'send messages to chat room' do
+      let(:message) { build(:message) }
+
+      subject { perform :send_message, message: message.body, chat_room_id: chat_room.id }
```
Comment (gabrielbursztein2) [inline spec/channels/chat_room_channel_spec.rb]: ()

File: spec/request/api/v1/chat_rooms/show_spec.rb [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+      get api_v1_chat_room_path(chat_room), headers: auth_header
+    end
+
+    it 'is expected a successful response' do
+      expect(response).to be_successful
+    end
+
+    it 'is expected that response contains some body data' do
+      body = (JSON response.body)
+      expect(json_value(body, 'chat_room', 'id')).not_to be_nil
```
Comment (gabrielbursztein2) [inline spec/request/api/v1/chat_rooms/show_spec.rb]: Use `include_json` matcher

File: spec/request/api/v1/targets/create_spec.rb [hunk 9]
```diff
@@ -64,6 +64,68 @@
...
+                country_code: 'US'
+              }
+            ]
+          )
+          new_target.save!
+        end
+
+        it 'is expected a successful response' do
+          subject
+          expect(response).to have_http_status(:ok)
```
Comment (gabrielbursztein2) [inline spec/request/api/v1/targets/create_spec.rb]: `be_successful`

File: spec/request/api/v1/targets/create_spec.rb [hunk 10]
```diff
@@ -64,6 +64,68 @@
...
+        end
+
+        it 'is expected the same count of delay_jobs' do
+          subject
+          expect { subject }.to change(Delayed::Job, :count).by(0)
```
Comment (gabrielbursztein2) [inline spec/request/api/v1/targets/create_spec.rb]: `.not_to change(Delayed::Job, :count)`

File: spec/request/api/v1/targets/create_spec.rb [hunk 11]
```diff
@@ -64,6 +64,68 @@
...
+        end
+
+        it 'is expected the same count of chat_rooms' do
+          subject
+          expect { subject }.to change(ChatRoom, :count).by(0)
```
Comment (gabrielbursztein2) [inline spec/request/api/v1/targets/create_spec.rb]: Same

File: app/channels/application_cable/chat_room_channel.rb (hunk 2 above)
Comment (afmicc) [reply app/channels/application_cable/chat_room_channel.rb:15]: Really, there's a database call in line `(18) message = current_user.messages.create!`. I want that if this call crash, don't send any notification

File: app/models/chat_room.rb (hunk 5 above)
Comment (JuanfraM) [reply app/models/chat_room.rb:15]: I know @gabrielbursztein2 that's a better approach but for the purpose of this induction, I think it's enough as it is.

File: app/controllers/api/v1/chat_rooms_controller.rb [hunk 12]
```diff
@@ -0,0 +1,13 @@
+module Api
+  module V1
+    class ChatRoomsController < ApiController
+      def index
+        @chat_rooms = current_user.chat_rooms
+      end
+
+      def show
+        @chat_room = current_user.chat_rooms.find(params[:id])
+      end
```
Comment (JuanfraM) [inline app/controllers/api/v1/chat_rooms_controller.rb:16]: you can def a function:
```
def chat_rooms
  current_user.chat_rooms
end
```
and use it above

File: app/models/target.rb [hunk 13]
```diff
@@ -63,9 +63,23 @@ def validate_target_limit
     errors.add(:targets, I18n.t('model.targets.errors.to_many'))
   end
 
-  def notify_compatible
-    users = near_targets.map(&:user).uniq(&:id)
+  def notify_and_create_chats
+    targets = near_targets
+    ActiveRecord::Base.transaction do
```
Comment (JuanfraM) [inline app/models/target.rb:57]: you should check and return if `targets` is empty. In this case, you avoid the transaction and also you can remove this precondition in the method `notify_compatible` (if you call it only here)

File: spec/request/api/v1/chat_rooms/index_spec.rb [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+describe 'List ChatRooms', type: :request do
+  let!(:user) { create(:user, :confirmed) }
+  let!(:user_guest) { create(:user, :confirmed) }
+  let!(:chat_room) do
+    create(:chat_room, :with_messages,
+           user_owner: user,
+           user_guest: user_guest)
+  end
+
+  describe 'GET api/v1/chat_rooms' do
```
Comment (JuanfraM) [inline spec/request/api/v1/chat_rooms/index_spec.rb]: Just this `GET api/v1/chat_rooms` should be the name of the describe. The first one is unnecesary

**Review by JuanfraM — APPROVED** [review]
Some comments.
Approved after fix it.
Good Job 💪

**Pull request closed without merging**